File: Python/0791.py
from collections import Counter

class Solution:
    def customSortString(self, order: str, s: str) -> str:
        ans = ""
        count = Counter(s)

        for cur in order:
            if cur in count:
                ans += cur * count[cur]
                count.pop(cur)
        
        for key, value in count.items():
            ans += key * value 

        return ans

# Topic: Hash Table, String
# Time Complexity: O(n)
# Space Complexity: O(n)

File: C++/0791.cpp
#include <string>
#include <unordered_map>

class Solution {
public:
    std::string customSortString(std::string order, std::string s) {
        std::string ans = "";
        std::unordered_map<int, int> count;

        for(auto& cur : s){
            count[cur]++;
        }
        for(auto& cur : order){
            if(count.find(cur) != count.end()){
                ans.append(count[cur], cur);
                count.erase(cur);
            }
        }
        for(auto& cur : count){
            ans.append(cur.second, cur.first);
        }
        return ans;
    }
};
